providers/huaweicloud/records.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return nil, err
+	}
+
+	for _, change := range changes {
+		msg := change.MsgsJoined
+		records := recordsToNative(change.New, change.Key)
+		rrsetsID := getRRSetIDFromRecords(change.Old)
+
+		switch change.Type {
+		case diff2.REPORT:
+			reports = append(reports, &models.Correction{Msg: msg})
+		case diff2.CREATE:
+			fallthrough
+		case diff2.CHANGE:
+			corrections = append(corrections, &models.Correction{
+				Msg: msg,
+				F: func() error {
+					if len(rrsetsID) == 1 {
+						return c.updateRRSet(zoneID, rrsetsID[0], records)
+					} else {
+						err := c.deleteRRSets(zoneID, rrsetsID)
+						if err != nil {
+							return err
+						}
+						return c.createRRSet(zoneID, records)
+					}
+				},
+			})
+		case diff2.DELETE:
+			deletions = append(deletions, &models.Correction{
+				Msg: msg,
+				F: func() error {
+					return c.deleteRRSets(zoneID, rrsetsID)
+				},
+			})
+		default:
+			panic(fmt.Sprintf("unhandled change.Type %s", change.Type))
+		}
+	}
+
+	result := append(reports, deletions...)
+	result = append(result, corrections...)
+	return result, nil
+}
+
+func (c *huaweicloudProvider) deleteRRSets(zoneID string, rrsets []string) error {
+	for _, rrset := range rrsets {
+		deletePayload := model.DeleteRecordSetRequest{
+			ZoneId:      zoneID,
+			RecordsetId: rrset,
+		}
+		var err error
+		withRetry(func() error {
+			_, err = c.client.DeleteRecordSet(&deletePayload)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *huaweicloudProvider) createRRSet(zoneID string, rc *model.ListRecordSets) error {
+	createPayload := &model.CreateRecordSetRequest{
+		ZoneId: zoneID,
+		Body: &model.CreateRecordSetRequestBody{
+			Name:    *rc.Name,
+			Type:    *rc.Type,
+			Ttl:     rc.Ttl,
+			Records: *rc.Records,
+		},
+	}
+	var err error
+	withRetry(func() error {
+		_, err = c.client.CreateRecordSet(createPayload)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *huaweicloudProvider) updateRRSet(zoneID, rrsetID string, rc *model.ListRecordSets) error {
+	updatePayload := &model.UpdateRecordSetRequest{
+		ZoneId:      zoneID,
+		RecordsetId: rrsetID,
+		Body: &model.UpdateRecordSetReq{
+			Name:    *rc.Name,
+			Type:    *rc.Type,
+			Ttl:     rc.Ttl,
+			Records: rc.Records,
+		},
+	}
+	var err error
+	withRetry(func() error {
+		_, err = c.client.UpdateRecordSet(updatePayload)
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func parseMarkerFromURL(link string) (string, error) {
+	// Parse the marker params from the URL
+	// Example: https://dns.myhuaweicloud.com/v2/zones?marker=abcdefg
+	url, err := url.Parse(link)
+	if err != nil {
+		return "", err
+	}
+	marker := url.Query().Get("marker")
+	if marker == "" {
+		return "", fmt.Errorf("marker not found in URL %s", link)
+	}
+	return marker, nil
+}
+
+func (c *huaweicloudProvider) getZoneRecords(zoneID string) (*[]model.ListRecordSets, error) {
```

Comment:
Please rename this function.  having GetZoneRecords() and getZoneRecords(), while valid in Go, could be confusing.

Thanks!